### PR TITLE
External Signature integration in core

### PR DESCRIPTION
**Pull request by adridadou**

### CODE REVIEW

File: shared/src/test/scala/org/adridadou/openlaw/vm/OpenlawVmSpec.scala [hunk 1]
```diff
@@ -87,6 +87,40 @@ class OpenlawVmSpec extends FlatSpec with Matchers {
...
+    )
+
+    val vm = vmProvider.create(definition, None, OpenlawSignatureOracle(TestCryptoService, serverAccount.address, Map(serviceName -> serviceAccount.address)), Seq())
+
+    vm(LoadTemplate(template))
+    vm.executionState shouldBe ContractCreated
+    vm.allNextActions.map(_.map(_.action)) shouldBe Right(Seq(SignatureAction(identity.email)))
+
+    val badSignature = EthereumSignature(signByEmail(identity.email, vm.contractId.data, serverAccount).signature)
+    val signature = EthereumSignature(signByEmail(identity.email, vm.contractId.data, serviceAccount).signature)
```
Comment (fforbeck) [inline shared/src/test/scala/org/adridadou/openlaw/vm/OpenlawVmSpec.scala]: What is the difference between the `badSignature` and `signature`?

Comment (adridadou) [reply shared/src/test/scala/org/adridadou/openlaw/vm/OpenlawVmSpec.scala]: it's the private key used to generate the signature. The idea is to show that if you try to sign with the wrong private key, it doesn't work

Comment (adridadou) [reply shared/src/test/scala/org/adridadou/openlaw/vm/OpenlawVmSpec.scala]: if you look closely, one is using serverAccount where the other uses serviceAccount

Comment (fforbeck) [reply shared/src/test/scala/org/adridadou/openlaw/vm/OpenlawVmSpec.scala]: Got it. Thanks!

File: shared/src/test/scala/org/adridadou/openlaw/vm/OpenlawVmSpec.scala [hunk 2]
```diff
@@ -87,6 +87,40 @@ class OpenlawVmSpec extends FlatSpec with Matchers {
     vm1.signature(identity2.email) shouldBe None
   }
 
+  it should "be possible to sign the contract with an external signature" in {
+    val template = """this is a contract [[Signatory:ExternalSignature(service:"DocuSign")]]"""
```
Comment (fforbeck) [inline shared/src/test/scala/org/adridadou/openlaw/vm/OpenlawVmSpec.scala]: Shouldn't we pass the `identity` argument to `ExternalSignature`?
Something like
```
[[identity: Identity]]
this is a contract [[Signatory:ExternalSignature(service:"DocuSign", identity: identity)]]
```

File: shared/src/test/scala/org/adridadou/openlaw/vm/OpenlawVmSpec.scala [hunk 3]
```diff
@@ -87,6 +87,40 @@ class OpenlawVmSpec extends FlatSpec with Matchers {
...
+    val templateId = TemplateId(TestCryptoService.sha256(template))
+
+    val serviceAccount = TestAccount.newRandom
+    val serviceName = ServiceName("DocuSign")
+
+    val definition = ContractDefinition(
+      creatorId = UserId.SYSTEM_ID,
+      mainTemplate = templateId,
+      templates = Map(),
+      parameters = TemplateParameters(Map(VariableName("Signatory") -> ExternalSignatureType.internalFormat(ExternalSignature(serviceName = serviceName, identity = Some(identity))).right.value))
```
Comment (fforbeck) [inline shared/src/test/scala/org/adridadou/openlaw/vm/OpenlawVmSpec.scala]: I noticed you passed the value as template parameter, but wasn't sure if we need that declared.

File: shared/src/main/scala/org/adridadou/openlaw/oracles/OpenlawSignatureProof.scala [hunk 4]
```diff
@@ -9,13 +9,25 @@ import io.circe.parser._
...
+  implicit val openlawSignatureProofEnc:Encoder[ExternalSignatureProof] = deriveEncoder
+  implicit val openlawSignatureProofDec:Decoder[ExternalSignatureProof] = deriveDecoder
+  def deserialize(json:String):Either[Error, ExternalSignatureProof] = decode[ExternalSignatureProof](json)
+}
+
 case class OpenlawSignatureProof(contractId:ContractId, fullName:String, signature:EthereumSignature, txHash:EthereumHash) extends SignatureProof {
-  override def serialize: String = this.asJson.noSpaces
+  override def serialize: Json = this.asJson
+
+  override def validationLink: Link = Link("verify signature",s"/signature/validate?contractId=${contractId.id}&signature=${signature.toString}")
```
Comment (outkaj) [inline shared/src/main/scala/org/adridadou/openlaw/oracles/OpenlawSignatureProof.scala]: i'm trying to understand if it's intentional that these two functions are declared twice?

File: shared/src/main/scala/org/adridadou/openlaw/oracles/StopContractOracle.scala [hunk 5]
```diff
@@ -1,57 +1,33 @@
...
 
 object StopExecutionEvent {
-  implicit val stopExecutionEventEnc:Encoder[StopExecutionEvent] = deriveEncoder[StopExecutionEvent]
-  implicit val stopExecutionEventDec:Decoder[StopExecutionEvent] = deriveDecoder[StopExecutionEvent]
+  implicit val stopExecutionEventEnc:Encoder[StopExecutionEvent] = deriveEncoder
+  implicit val stopExecutionEventDec:Decoder[StopExecutionEvent] = deriveDecoder
 }
 
-case class StopExecutionEvent(signature: EthereumSignature) extends OpenlawVmEvent {
+case class StopExecutionEvent(signature: EthereumSignature) extends SignedActionEvent {
```
Comment (outkaj) [inline shared/src/main/scala/org/adridadou/openlaw/oracles/StopContractOracle.scala]: we should check on the backend to make sure stop / resume still work with these changes, I know there have been some bugs with those features in the past

File: shared/src/main/scala/org/adridadou/openlaw/oracles/OpenlawSignatureProof.scala (hunk 4 above)
Comment (outkaj) [reply shared/src/main/scala/org/adridadou/openlaw/oracles/OpenlawSignatureProof.scala]: oh wait, never mind, now the file expanded and I see what's going on

**Pull request merged**